include/rigtorp/MPMCQueue.h

```diff
@@ -253,6 +253,21 @@ class Queue {
     }
   }
 
+  /// Returns the number of elements in the queue.
+  /// The size can be negative when the queue is empty and there is at least one
+  /// reader waiting. Since this is a concurrent queue the size is only a best
+  /// effort guess until all reader and writer threads have been joined.
+  ptrdiff_t size() const noexcept {
```

Comment:
FYI, in my versions, the implementation is below
`
  size_t size() const noexcept {
    std::ptrdiff_t diff = head_.load(std::memory_order_acquire) -
                          tail_.load(std::memory_order_acquire);
    if (diff < 0) {
      diff += capacity_;
    }
    return static_cast<size_t>(diff);
  }
`

Reply:
The name size() is misleading for such a queue. I have seen names with better semantics such as:
was_size() or volatile_size()

Reply:
> 
I think the size method in a mpmc queue class already implies that, anyone should have known that before using it.

Reply:
> I think the size method in a mpmc queue class already implies that, anyone should have known that before using it.

Yes, you need to know the implementation, and that's precisely the problem. If someone in the future needs to maintain some code using this queue, he could be misled.

Reply:
> FYI, in my versions, the implementation is below
> 
> ```
> 
>   size_t size() const noexcept {
>     std::ptrdiff_t diff = head_.load(std::memory_order_acquire) -
>                           tail_.load(std::memory_order_acquire);
>     if (diff < 0) {
>       diff += capacity_;
>     }
>     return static_cast<size_t>(diff);
>   }
> ```

That's not quite right, if you are using the blocking pop() you can have `diff < 0`, it would be negative the number of waiting readers. The `size()` implementation is tricky to get right for the case of overflowed head or tail. One approach is to make them always be uint64_t even on 32bit arch and assume no overflow. But your code is definitely wrong.

Reply:
> > I think the size method in a mpmc queue class already implies that, anyone should have known that before using it.
> 
> Yes, you need to know the implementation, and that's precisely the problem. If someone in the future needs to maintain some code using this queue, he could be misled.

If you're building a multi-threaded application and you don't understand that `size()` can change concurrently while reading it, that's your problem, not the name of the method. I specifically mention that `size()` is not a synchronization point, only after all threads reading or writing to the queue has been joined, can you rely on the value of `size()`